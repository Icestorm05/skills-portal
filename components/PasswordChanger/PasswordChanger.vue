<template lang="pug">
    v-form#PasswordChanger(@submit.prevent="submit")
        v-text-field(:append-icon="oldPasswordVisible ? 'visibility_off' : 'visibility'"
                     :append-icon-cb="() => oldPasswordVisible = !oldPasswordVisible"
                     autocomplete="new-password"
                     auto-grow
                     label="Old Password"
                     name="old-password"
                     hint="Enter your old password here"
                     :type="oldPasswordVisible ? 'text' : 'password'"
                     v-model="oldPassword"
                     v-validate="'required|old-password-match'"
                     data-vv-name="old-password"
                     data-vv-as="Old Password"
                     :data-vv-delay="oldPasswordCheck ? oldPasswordCheckDelay : 0"
                     :error-messages="errors.collect('old-password')"
                     :loading="checkingOldPassword"
                     :disabled="disabled")#OldPassword
        v-text-field(:append-icon="newPasswordVisible ? 'visibility_off' : 'visibility'"
                     :append-icon-cb="() => newPasswordVisible = !newPasswordVisible"
                     autocomplete="new-password"
                     auto-grow
                     label="New Password"
                     name="new-password"
                     hint="Enter your new password here"
                     :type="newPasswordVisible ? 'text' : 'password'"
                     v-model="newPassword"
                     v-validate="'required|confirmed:confirmed-new-password'"
                     data-vv-name="new-password"
                     data-vv-as="New Password"
                     :error-messages="errors.collect('new-password')"
                     :disabled="disabled")#NewPassword
        v-text-field(:append-icon="confirmedNewPasswordVisible ? 'visibility_off' : 'visibility'"
                     :append-icon-cb="() => confirmedNewPasswordVisible = !confirmedNewPasswordVisible"
                     autocomplete="new-password"
                     auto-grow
                     label="Confirm New Password"
                     name="confirmed-new-password"
                     hint="Confirm your new password here"
                     :type="confirmedNewPasswordVisible ? 'text' : 'password'"
                     v-model="confirmedNewPassword"
                     v-validate="'required|confirmed:new-password'"
                     data-vv-name="confirmed-new-password"
                     data-vv-as="Confirmed New Password"
                     :error-messages="errors.collect('confirmed-new-password')"
                     :disabled="disabled")#ConfirmedNewPassword
        v-btn(color="secondary"
              raised
              ripple
              type="submit"
              :disabled="disabled")#PasswordChangerSubmitButton Submit
        v-btn(color="secondary"
              raised
              ripple
              @click="reset")#PasswordChangerResetButton Reset
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import VBtn from "vuetify/es5/components/VBtn";
import VForm from "vuetify/es5/components/VForm";
import VTextField from "vuetify/es5/components/VTextField";

@Component({
  components: {
    VBtn,
    VForm,
    VTextField
  },
  $_veeValidate: {
    validator: "new"
  }
})
export default class PasswordChanger extends Vue {
  @Prop({ type: Boolean, default: false })
  private disabled: boolean;
  @Prop({ type: Function })
  private oldPasswordCheck: ((password: string) => any) | undefined;
  @Prop({ type: Number, default: 500 })
  private oldPasswordCheckDelay: number;

  private oldPassword = "";
  private oldPasswordVisible = false;
  private checkingOldPassword = false;

  private newPassword = "";
  private newPasswordVisible = false;

  private confirmedNewPassword = "";
  private confirmedNewPasswordVisible = false;

  private mounted() {
    this.$validator.extend("old-password-match", {
      getMessage: field => {
        return `The ${field} does not match current password.`;
      },
      validate: async value => {
        if (!this.oldPasswordCheck) {
          return true;
        }
        try {
          this.checkingOldPassword = true;
          const match = await this.oldPasswordCheck(value);
          return match ? true : false;
        } catch (err) {
          return false;
        } finally {
          this.checkingOldPassword = false;
        }
      }
    });
  }

  private reset() {
    this.oldPassword = "";
    this.newPassword = "";
    this.confirmedNewPassword = "";
  }

  private async submit() {
    if (!this.disabled && await this.$validator.validateAll()) {
      this.$emit("password-change", this.oldPassword, this.newPassword);
    }
  }
}
</script>
