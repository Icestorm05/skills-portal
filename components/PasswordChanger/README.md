# Password Changer

## Functionality

The password changer is a form which allows a user to change their password. The form requires a user to enter their old password, as well as their new password.

### Props

`<password-changer>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `disabled` | Boolean *optional* | Whether or not the input elements are disabled or not. Defaults to `false`.
| `old-password-check` | Function *optional* | An optional function to determine whether or not the old password is valid or not. Takes a password argument, and should return a boolean. Can be async.
| `old-password-check-delay` | Number *optional* | How much input delay there is before executing the `old-password-check` function (in milliseconds). Defaults to `500`.

### Example 1 - Synchronous Validator

```html
<password-changer :old-password-check="passwordValidator"
                  :old-password-check-delay="1000">
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import PasswordChanger from '~/components/PasswordChanger/PasswordChanger.vue';
@Component({
    components: {
        PasswordChanger,
    },
})
export default class GenericComponent extends Vue {

    // Synchronous Validator
    private passwordValidator(password: string) {
        return password === 'test123';
    }

}
```

### Example 2 - Asynchronous Validator

```html
<password-changer :old-password-check="passwordValidator">
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import PasswordChanger from '~/components/PasswordChanger/PasswordChanger.vue';
@Component({
    components: {
        PasswordChanger,
    },
})
export default class GenericComponent extends Vue {

    // Asynchronous Validator
    private passwordValidator(password: string) {
        return new Promise(resolve => {
            window.setTimeout(() => {
                if (password === 'test123') {
                    resolve(true); // Validates it
                } else {
                    resolve(false); // Invalidates it
                    //reject(); // Also works
                }
            }, 500);
        });
    }

}
```