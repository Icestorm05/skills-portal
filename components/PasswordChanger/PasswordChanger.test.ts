import PasswordChanger from './PasswordChanger.vue';

import VeeValidate from 'vee-validate';
import { createLocalVue, mount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(VeeValidate);

declare const expect: jest.Expect;

describe('Password Changer', () => {

    it('is a Vue instance', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('snapshot has not changed', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('disabled prop disables all inputs', () => {
        const wrapper = mount(PasswordChanger, { localVue, propsData: { disabled: true } });
        expect(wrapper.find('#PasswordChanger #OldPassword').attributes().disabled).toEqual('disabled');
        expect(wrapper.find('#PasswordChanger #NewPassword').attributes().disabled).toEqual('disabled');
        expect(wrapper.find('#PasswordChanger #ConfirmedNewPassword').attributes().disabled).toEqual('disabled');
    });

    it('entering into old password input updates old password data', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        const oldPassword = wrapper.find('#PasswordChanger #OldPassword');
        (oldPassword.element as HTMLInputElement).value = 'test123';
        oldPassword.trigger('input');
        expect(wrapper.vm.$data.oldPassword).toEqual('test123');
    });

    it('entering into new password input updates new password data', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        const newPassword = wrapper.find('#PasswordChanger #NewPassword');
        (newPassword.element as HTMLInputElement).value = 'test123';
        newPassword.trigger('input');
        expect(wrapper.vm.$data.newPassword).toEqual('test123');
    });

    it('entering into confirm new password input updates confirm new password data', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        const confirmedNewPassword = wrapper.find('#PasswordChanger #ConfirmedNewPassword');
        (confirmedNewPassword.element as HTMLInputElement).value = 'test123';
        confirmedNewPassword.trigger('input');
        expect(wrapper.vm.$data.confirmedNewPassword).toEqual('test123');
    });

    it('form submit with invalid inputs does not emit the password-change event', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted('password-change')).toBeFalsy();
    });

    it('form submit with valid inputs emits the password-change event', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ oldPassword: 'test123', newPassword: 'test321', confirmedNewPassword: 'test321' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted('password-change')).toEqual([['test123', 'test321']]);
    });

    it('form submit with valid inputs and disabled prop does not emit the password-change event', async() => {
        const wrapper = mount(PasswordChanger, { localVue, propsData: { disabled: true } });
        wrapper.setData({ oldPassword: 'test123', newPassword: 'test321', confirmedNewPassword: 'test321' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.emitted('password-change')).toBeFalsy();
    });

    it('reset button click resets all inputs', () => {
        const wrapper = mount(PasswordChanger, { localVue, propsData: { disabled: true } });
        wrapper.setData({ oldPassword: 'test123', newPassword: 'test321', confirmedNewPassword: 'test321' });
        wrapper.find('#PasswordChanger #PasswordChangerResetButton').trigger('click');
        expect(wrapper.vm.$data.oldPassword).toEqual('');
        expect(wrapper.vm.$data.newPassword).toEqual('');
        expect(wrapper.vm.$data.confirmedNewPassword).toEqual('');
    });

    it('old password is invalid with nothing entered', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ oldPassword: '' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toEqual('The Old Password field is required.');
    });

    it('new password is invalid with nothing entered', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ newPassword: '' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('new-password')).toEqual('The New Password field is required.');
    });

    it('confirmed new password is invalid with nothing entered', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ confirmedNewPassword: '' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('confirmed-new-password')).toEqual('The Confirmed New Password field is required.');
    });

    it('new password is invalid with mismatching confirmed new password', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ newPassword: 'test321', confirmedNewPassword: 'test123' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('new-password')).toEqual('The New Password confirmation does not match.');
    });

    it('confirmed new password is invalid with mismatching new password', async() => {
        const wrapper = mount(PasswordChanger, { localVue });
        wrapper.setData({ newPassword: 'test321', confirmedNewPassword: 'test123' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('confirmed-new-password')).toEqual('The Confirmed New Password confirmation does not match.');
    });

    it('adds a custom old-password-match validator to the old-password input', () => {
        const wrapper = mount(PasswordChanger, { localVue });
        const oldPassword = wrapper.vm.$validator.fields.find({ name: 'old-password' });
        expect(Object.keys(oldPassword.rules).some(key => key === 'old-password-match')).toBeTruthy();
    });

    it('old password is invalid with failing validation fn', async() => {
        const oldPasswordCheck = (value: string) => value === 'test123';
        const wrapper = mount(PasswordChanger, { localVue, propsData: { oldPasswordCheck } });
        wrapper.setData({ oldPassword: 'test321' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toEqual('The Old Password does not match current password.');
    });

    it('old password is valid with correct validation fn', async() => {
        const oldPasswordCheck = (value: string) => value === 'test123';
        const wrapper = mount(PasswordChanger, { localVue, propsData: { oldPasswordCheck } });
        wrapper.setData({ oldPassword: 'test123' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toBeNull();
    });

    it('old password is invalid with failing resolved async validation fn', async() => {
        const oldPasswordCheck = (value: string) => new Promise(resolve => {
            resolve(value === 'test123');
        });
        const wrapper = mount(PasswordChanger, { localVue, propsData: { oldPasswordCheck } });
        wrapper.setData({ oldPassword: 'test321' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toEqual('The Old Password does not match current password.');
    });

    it('old password is invalid with failing rejected async validation fn', async() => {
        const oldPasswordCheck = (value: string) => new Promise((resolve, reject) => {
            return value === 'test123' ? resolve(true) : reject();
        });
        const wrapper = mount(PasswordChanger, { localVue, propsData: { oldPasswordCheck } });
        wrapper.setData({ oldPassword: 'test321' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toEqual('The Old Password does not match current password.');
    });

    it('old password is valid with passing async validation fn', async() => {
        const oldPasswordCheck = (value: string) => new Promise((resolve, reject) => {
            return value === 'test123' ? resolve(true) : reject();
        });
        const wrapper = mount(PasswordChanger, { localVue, propsData: { oldPasswordCheck } });
        wrapper.setData({ oldPassword: 'test123' });
        wrapper.find('#PasswordChanger').trigger('submit');
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.$validator.errors.first('old-password')).toBeNull();
    });

});
