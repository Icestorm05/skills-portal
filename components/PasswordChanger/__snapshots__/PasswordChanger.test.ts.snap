// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Changer snapshot has not changed 1`] = `
<form novalidate="novalidate" id="PasswordChanger">
    <div class="input-group input-group--append-icon input-group--text-field primary--text" data-vv-id="6" aria-required="true" aria-invalid="false">
        <label for="OldPassword">Old Password</label>
        <div class="input-group__input">
            <input autocomplete="new-password" name="old-password" data-vv-name="old-password" data-vv-as="Old Password" data-vv-delay="0" id="OldPassword" tabindex="0" type="password">
            <i aria-hidden="true" class="icon material-icons input-group__append-icon input-group__icon-cb">visibility</i>
        </div>
        <div class="input-group__details">
            <!---->
        </div>
    </div>
    <div class="input-group input-group--append-icon input-group--text-field primary--text" data-vv-id="7" aria-required="true" aria-invalid="false">
        <label for="NewPassword">New Password</label>
        <div class="input-group__input">
            <input autocomplete="new-password" name="new-password" data-vv-name="new-password" data-vv-as="New Password" id="NewPassword" tabindex="0" type="password">
            <i aria-hidden="true" class="icon material-icons input-group__append-icon input-group__icon-cb">visibility</i>
        </div>
        <div class="input-group__details">
            <!---->
        </div>
    </div>
    <div class="input-group input-group--append-icon input-group--text-field primary--text" data-vv-id="8" aria-required="true" aria-invalid="false">
        <label for="ConfirmedNewPassword">Confirm New Password</label>
        <div class="input-group__input">
            <input autocomplete="new-password" name="confirmed-new-password" data-vv-name="confirmed-new-password" data-vv-as="Confirmed New Password" id="ConfirmedNewPassword" tabindex="0" type="password">
            <i aria-hidden="true" class="icon material-icons input-group__append-icon input-group__icon-cb">visibility</i>
        </div>
        <div class="input-group__details">
            <!---->
        </div>
    </div>
    <button type="submit" class="btn secondary" raised="raised" id="PasswordChangerSubmitButton">
        <div class="btn__content">Submit</div>
    </button>
    <button type="button" class="btn secondary" raised="raised" id="PasswordChangerResetButton">
        <div class="btn__content">Reset</div>
    </button>
</form>
`;
